src/flb_log.c

```diff
@@ -62,10 +62,83 @@ static inline int consume_byte(flb_pipefd_t fd)
     return 0;
 }
 
+static inline int log_file_sz_max(off_t size, struct flb_log *log)
+{
+    double sz_mb;
+
+    sz_mb = size / (1024.00 * 1024.00);
+    if (sz_mb >= log->max_sz_mb) {
+        return FLB_TRUE;
+    }
+
+    return FLB_FALSE;
+}
+
+static inline void log_file_rotate(struct flb_log *log)
+{
+    char *backup_file_name = NULL;
+    char *src_file_name = NULL;
+    int ret = -1;
+    int i;
+
+    if (log->max_history == 0) {
+        /*
+         * no backups.
+         * remove the file since size has exceeded.
+         */
+        remove(log->out);
+        return;
+    }
+
+    /* max 2 digits for backup name suffix and 1 each for '.' and '\0' */
+    size_t filename_len = strlen(log->out) + 4;
+    backup_file_name = flb_malloc(filename_len);
+    if (!backup_file_name) {
+        flb_errno();
+        return;
+    }
+    src_file_name = flb_malloc(filename_len);
+    if (!src_file_name) {
+        flb_errno();
+        return;
+    }
+
+    for (i = log->max_history; i > 1; i--) {
+        snprintf(backup_file_name, filename_len, "%s.%d", log->out, i);
+        snprintf(src_file_name, filename_len, "%s.%d", log->out, i-1);
+        /* since rename behaviour is implementation dependent when new file exists, remove the last file. */
+        if (i == log->max_history) {
+            remove(backup_file_name);
+        }
+        ret = rename(src_file_name, backup_file_name);
+        if (ret != 0) {
+            /* we could not rename; delete the source file. */
+            remove(src_file_name);
+        }
+    }
+    if (src_file_name) {
```

Comment:
Can we move this block to line 130 and remove the if blocks? if we got there it means the pointers are or at the very least should be, we could check them and report if they are not but I think it might not really be all that useful.